New tool to prepare ginan to run using a rinex file as input

aclel opened the pull request:
This PR proposes to introduce a new tool to help users when running Ginan with their own RINEX files for the first time.

* It parses the RINEX file to extract information from the header and observations
* It downloads the required IGS products to run Ginan for a ppp static solution
* It generates a yaml config for Ginan to be able to get a ppp static solution

These three functions are unified in a `prep` command. This command currently has a narrow use case - to allow Ginan to run RINEX files that would normally be submitted to state government for inclusion in the national adjustment. The intention is to enhance the tool's capabilities over time to support more use cases.

Please suggest a better name than `gn` - I wanted something that wasn't `ginan` and it's just what came to mind. Maybe there is an aboriginal word in the same realm as `ginan` that would have some significance?

aclel committed:
Download deps required to run pea for a RINEX file

- Download model dependencies (ATX, BLQ, Trop)

- Download products specific to the RINEX file (ERP, SP3)

aclel committed:
Fix CLK download

aclel committed:
Fix filenames when downloading from CDDIS FTP

- Product files are stored as .gz

- TODO to look at gnssanalysis download. It just creates an empty file
if the file does not exist

aclel committed:
Add script to call yaml generator

aclel committed:
Write config to "network" directory

aclel committed:
Add yaml template

aclel committed:
Use black formatter, use spaces

aclel committed:
Change ModelDependencyDownloader to functions

- Function was not necessary - it just wrapped around one argument

aclel committed:
Add missing argument to download functions

aclel committed:
Add script to download rinex deps and build yaml

- Only works for rinex v3

aclel committed:
Delete model dependency downloader

- This can just be done by the s3_filehandler

aclel committed:
Merge remote-tracking branch 'upstream/develop-weekly' into rinex-single-station

aclel committed:
Download requires products for all gps days in a rinex file

- A RINEX file can run across multiple GPS days, and an orbit file is
required for each GPS day

aclel committed:
Only write overrides to the yaml template

- auto_generate_yaml contains functions which allow you to specify
particular files which should be included in the yaml. Globs can be used
instead of specifying files.

- All that needs to be changed in the template at the moment is the
station options, and the config name

- This will likely change to be more complex in future

aclel committed:
Change sinex output path to just the config name for now

aclel committed:
Use proc_noise of 0 for static solution

- proc_noise of 100 is used for a kinematic solution

aclel committed:
Clean up template naming

aclel committed:
Add cli interface for building config

- Copies the rinex file into a new config directory

- Generates the yaml and stores it in the same directory

aclel committed:
Parse code priorities from rinex observations

- The rinex header does not reliably give the signals that are tracked
by the receiver for each system. Instead the observations are parsed
with georinex and then interrogated.

- Code priorities are then used to populate the yaml

aclel committed:
Rename rinex file to have .rnx extension on copy

- When the rinex file is copied into the workspace it needs to be
renamed so that the *.rnx glob in the template can match the file

aclel committed:
Rename stations to RXXX, where XXX are last 3 chars of marker name

- pea does not support station names that are all numbers

- reformat the name to a string with R prefix to force it to work

aclel committed:
Enable phase bias options and change bias file extension

aclel committed:
Update config template for ginan v3

- Config file has been restructured

- Also fix eccentricity format - needs to be ENU for Ginan, which is
different to UEN in the SINEX

aclel committed:
Create classes for RinexHeader

- Use typing instead of a dict

aclel committed:
Use gnssanalysis for download - move some functions there

- gnssanalysis.download is the point of truth for downloading from
cddis. Now using these functions instead of writing our own.

- auto_download_PPP had some functions which made more sense to have in
gnssanalysis.filenames - these have been moved out and imported instead

aclel committed:
Move ftp_tls to gnssanalysis and use in download script

aclel committed:
Move yaml overrides to their own function

aclel committed:
Merge remote-tracking branch 'upstream/develop-weekly' into rinex-single-station

aclel committed:
Move download functions to gnssanalysis

aclel committed:
Move long filename cutoff back to auto_download_PPP

- Had moved to gnssanalysis but don't want to fry that fish yet

aclel committed:
Use newer download functions rather than download_prod

- download_prod will likely be deprecated as it handles only handles
short filenames. Work will need to be done to eventually support short
and long filenames.

- this is a simple change to download just what is necessary for static
ppp of a single rinex file with a IGS FIN products - nothing more

aclel committed:
Move scripts into gn module

- Can be used to create nice cli tools around the outside of ginan

aclel committed:
Change cli to use flags for ppp static

Instead of grouping ppp, it makes more sense to have it as a flag. In
future it might make sense to incorporate orbit determination flags etc
as well.

aclel committed:
Set up logging

aclel committed:
Add README

aclel committed:
Update instructions with pea command and folders

aclel committed:
Rename config-name to workspace-name, More instructions in readme

aclel committed:
Merge remote-tracking branch 'upstream/develop-weekly' into rinex-single-station

aclel committed:
Reset auto_download_PPP back to upstream

aclel committed:
Run black auto formatter on auto_download_PPP

aclel committed:
Revert "Run black auto formatter on auto_download_PPP"

This reverts commit b8e9d86390640be73b60bb5f50d190e0666042b3.

aclel committed:
Reset auto_download_PPP back to develop-weekly upstream

aclel committed:
Remove whitespace at the end of file

Reset to upstream version

ronaldmaj commented on scripts/auto_generate_yaml.py:
Just checking, you didn't end up using this file in your implementation right?

aclel replied:
Mostly no - I found by creating the template yaml all I really needed was the ability to override options in the file itself, rather than generate the file from scratch. I copied the write_nested_dict_value function into my tool. I wanted to come up with a better way - any suggestions? Could import from auto_generate_yaml.py for now?

ronaldmaj replied:
Nah, not a worry. This was a pretty old script so hadn't been updated in a while. I was contemplating starting from scratch anyway, so for now you can leave the function you took out in your scripts

ronaldmaj commented on scripts/gn/README.md:
```diff
@@ -0,0 +1,44 @@
+
+## Getting started
+
+Ensure you have created a virtual environment in the `scripts` folder and installed the `requirements.txt`.
```
We could include instructions on how to do this via venv for example (or point people to the relevant docs that explain how to set up a virtual env)

aclel replied:
Just added some instructions to the README

ronaldmaj commented on scripts/gn/README.md:
```diff
@@ -0,0 +1,44 @@
+
+Ensure you have created a virtual environment in the `scripts` folder and installed the `requirements.txt`.
+
+Also ensure you have run `python3 scripts/s3_filehandler.py -d -p` to populate the `inputData` folder with the constants that `pea` depends on.
```
I would like to replace this with the `auto_download_PPP` script. I think I've got most (all?) of the relevant inputs now covered, especially following the new ver 3 ppp_example template. 
I think for now this is ok, but we'll need to do some testing seeing how that script can be incorporated here

aclel replied:
Awesome, I'll take a look at `auto_download_PPP` and we can look at incorporating it in near future

ronaldmaj replied:
Actually, I realise now that it hasn't entered the GitHub repo yet, so maybe I'll send you a copy of that now and you can have a look at it from that version

ronaldmaj commented on scripts/gn/download_rinex_deps.py:
Some of this is now covered in the `auto_download_PPP` script, so we'll need to look at merging what is done here with the use of that script (or at least some of the functions there)
Again, what you've got here is good, and I think it's fine to go in, but we'll need to have a closer look at that in the next step

aclel replied:
Sounds good - next step is to merge and refine

ronaldmaj commented on scripts/gn/parse_rinex_header.py:
This is really awesome and something we haven't had in Python form yet - this will bring a lot of value for people using Ginan with their own RINEX file

ronaldmaj reviewed (commented):
This is looking really great! I still need to test it out, but from what I've seen it's probably in good enough shape to go in. I might get a few more eyeballs on this as well before then

aclel committed:
Add venv instructions to readme

treefern commented on scripts/gn/README.md:
```diff
@@ -0,0 +1,54 @@
+* Only works on RINEX files collected after 2022-11-27 (long filenames started on CDDIS)
+* Only handles a single RINEX file at a time
+* Only handle downloads from CDDIS
+* Recommendation to use RINEX files with >6 hours of data. The generated yaml configuration has ambiguity resolution is turned off, so while pea will still run with a shorter observation session, the results may be well off.
```
```suggestion
* Recommendation to use RINEX files with >6 hours of data. The generated yaml configuration has ambiguity resolution turned off, so while pea will still run with a shorter observation session, the results may be much less accurate.
```
Small typo fix. Change of phrasing to avoid using an idiom for `inaccurate` which may confuse some readers.

treefern commented on scripts/gn/main.py:
```diff
@@ -0,0 +1,209 @@
+        return yaml_path
+    else:
+        raise NotImplementedError(
+            "Only static ppp is supported through this interface - include --static --ppp"
```
```suggestion
            "Only static ppp is supported through this interface - include --static --ppp "
```
Will concatenate to `-pppYou` as is

treefern commented on scripts/gn/parse_rinex_header.py:
```diff
@@ -0,0 +1,186 @@
+
+    marker_name = header["MARKER NAME"].strip()
+
+    # TODO: Handle the case when not all the fields are given - can't unpack 3 fields when there are only two values
```
I appreciate the call out on this one, given issues I've seen before.
Can we determine in a straightforward way whether the fields are missing, and raise an exception on it?

aclel replied:
It's hard to tell which fields are missing. It's possible that only the TYPE (middle value) could be provided, and when we split we would only get one value back. It would be hard to tell that we're missing "REC #" and "VERS".

We would need some kind of way to check that each element matches a pattern, ie. check if the TYPE is in a list of valid receivers, does the REC # conform to a certain pattern, is the VERS valid. These would be assumptions and I'm not sure if user data would always follow these patterns.

I'm not sure the best way of handling it at the moment.

treefern replied:
Right. Yeah hard to handle if less than three values **could be valid**, but telling what is what, and validating it is non-trivial.

I guess !=3 elements: raise `NotImplemented` for now? (Are we actually looking at values though, or just the header?)

I can't quite see how the `TYPE` is being _extracted_ as it is on line 82.
~~Line 69 is also confusing me... ohh~~. I'd suggest fully qualifying `gr` as `georinex`. I misread it as `gn` several times and couldn't figure how anything was working given `rinexheader()` isn't defined locally, but there's a class called `RinexHeader` with no explicit constructor.

aclel replied:
Line 82 was what I was using to parse a specific file that I have that only has TYPE (not the other two elements), but it needs to be more general.

The only element that's being used is TYPE (receiver type), which ends up in `receiver_options: STATION: receiver_type:` in the ginan yaml. Does ginan actually use this information for anything? Or does it just care about the *antenna type* to pull out the correct values from the igs20.atx file?

If it's not using it, then perhaps we don't parse it at all for now? If it is, then could we split the field into its elements and check any of the elements are found in a list of valid TYPEs (not sure where to get this list but I'm sure it would exist somewhere).

It would be nice if we could raise an exception if there are not 3 elements, but I've come across a fair few files that don't have this field fully populated. Trying our best to pull out a valid type might be nicer than asking the user to pad out the field with dummy data so that there are 3 elements to parse.

treefern replied:
Apparently `receiver_type` hardly gets used. And I haven't found a list of valid `TYPE`s anywhere.
So given it's also hard to parse out robustly, I'd agree we probably shouldn't parse it for now.

aclel replied:
See [4cedd91](https://github.com/GeoscienceAustralia/ginan/pull/62/commits/4cedd911bd985f36f8690452dbcca2a10d8f0941)

aclel committed:
Fix exception text

Co-authored-by: Nathan <95725385+treefern@users.noreply.github.com>

aclel committed:
Improve wording in readme

Co-authored-by: Nathan <95725385+treefern@users.noreply.github.com>

treefern commented on scripts/requirements.txt:
```diff
@@ -20,4 +20,6 @@ pymongo>=3.7.2
 pytest>=7.1.2
 scipy>=1.7.3
 statsmodels>=0.13.2
-gnssanalysis>=0.0.28
```
Was this intentional? `gnssanalysis` is leveraged in `download_rinex_deps.py`

aclel replied:
Whoops, not intentional. I had replaced it with my local version of gnssanalysis because I made some changes and they weren't merged in yet. Meant to leave it unchanged in requirements.

treefern commented on scripts/requirements.txt:
```diff
@@ -20,4 +20,6 @@ pymongo>=3.7.2
-gnssanalysis>=0.0.28
+#georinex>=1.16.2
+git+https://github.com/breid-phys/georinex.git # fast processing patch - not yet merged into georinex https://github.com/geospace-code/georinex/pull/85
+-e . # Add local gn
```
Is this here as a convenience primarily?
I gather it's streamlining things for a better user experience (same effect as running `pip install .` after gathering the above requirements?)

aclel replied:
Yep just for convenience

treefern replied:
After some other discussions I realised `-e` apparently _symlinks_, not installs. Still convenient, but maybe a bit of unexpected behaviour we should avoid doing in `requirements.txt`?

aclel replied:
Agree, see commit [998bc72](https://github.com/GeoscienceAustralia/ginan/pull/62/commits/998bc72b9fac517f391f52fd7fe63c2e3bb5cc46) which removes it from `requirements.txt` and adds a new instruction to the README to do a `pip install .`

aclel committed:
Re-add gnssanalysis to requirements.txt

Removal was unintentional

aclel committed:
Merge branch 'rinex-single-station' of github.com:aclel/ginan into rinex-single-station

aclel committed:
Change georinex import from gr

Use fully qualified name to avoid confusion with "gn"

aclel committed:
Remove gn from requirements.txt

"-e ." creates a symlink which is convenient for development, but it is
unexpected behaviour that could cause issues down the track. For
simplicity, ask the user to install with "pip install ."

treefern commented on scripts/gn/README.md:
```diff
@@ -12,8 +12,14 @@ cd /path/to/ginan/scripts
 python -m venv env
 
+# Activate the virtualenv
+. env/bin/activate
```
Valid, though `source` is _reasonably_ universal and probably clearer.

aclel replied:
Sounds good

aclel committed:
Use source to activate virtualenv

Clearer than "."

aclel committed:
Do not parse receiver from rinex

There is not currently a robust way to parse out the different
components of the receiver (num, type and vers). In future it may be
required, but for now ginan does not use it much, so we can safely
exclude it.

aclel committed:
Include yaml template in gn package

The switch from an explicit "-e ." in requirements.txt to "pip install
." caused an issue with the template file no longer being found. This
change installs the template alongside the package.

treefern commented:
Great stuff. 😄
I think the only remaining tweak from my perspective is whether some of the more general utility functions (like the Rinex header parser) move into `gnssanalysis`, but that can happen later on.

aclel commented:
Thanks for the feedback!😃 yeah keen to keep improving it over time. Moving some stuff into gnssanalysis sounds good

aclel committed:
Read start and end datetimes from obs timeseries instead of header

- The header does not have an end date in CORS files, because these
stations do not stop logging. We need to know the end date in the file
so that we know which dates to download products for, so it's more
robust to use the min and max datetimes from the observations themselves

aclel committed:
Parse antenna type if a number exists as well

Some rinex files include a number in the same row as antenna type. The
two parts need to be parsed in this situation.

ronaldmaj commented on scripts/gn/templates/auto_template.yaml:
```diff
@@ -0,0 +1,319 @@
+
+    atx_files:  [ ../../inputData/products/igs20.atx                    ] # required
+    egm_files:  [ ../../inputData/products/tables/EGM2008.gfc           ] # Earth gravity model coefficients file
+    igrf_files: [ ../../inputData/products/igrf13coeffs.txt             ]
```
Should be `../../inputData/products/tables/igrf13coeffs.txt` (i.e. in the tables directory)

aclel committed:
Fix ref to igrf13coeffs.txt in template yaml

igrf13coeffs.txt is downloaded to products/tables

ronaldmaj commented on scripts/gn/templates/auto_template.yaml:
```diff
@@ -0,0 +1,319 @@
+
+    outputs_root:              ./outputs/
+
+    # gpx:
```
It would be good to have gpx turned on by default - it's an easier way for users to grab results after the run. The TRACE is ok, but it's a big file and it can be hard to find the details you want
I think this will be my last comment on this (happy for it to go in after this).

ronaldmaj reviewed (changes requested):
I've completed the run and this works well! :+1:  
Would be nice not to have to rely on the `s3_filehandler` download as it comes along with a whole bunch of other files, but I think that'll be a point for the next iteration on this.

aclel committed:
Enable gpx output

The trace file is ok, but it's a big file and it's hard to find the
details you want. GPX file is an easy way for users to grab results
after the run.

ronaldmaj reviewed (approved):
Looks good now, I think all issues were addressed
